src/Symfony/Component/Security/Http/Firewall/AccessListener.php

```diff
@@ -30,22 +31,85 @@ class AccessListener implements ListenerInterface
     private $accessDecisionManager;
     private $map;
     private $authManager;
+    private $authorizationChecker;
 
-    public function __construct(TokenStorageInterface $tokenStorage, AccessDecisionManagerInterface $accessDecisionManager, AccessMapInterface $map, AuthenticationManagerInterface $authManager)
+    /**
+     * @param TokenStorageInterface|AuthorizationCheckerInterface $tokenStorage
+     * @param AccessDecisionManagerInterface|AccessMapInterface   $accessDecisionManager
+     * @param AccessMapInterface|null                             $map
+     * @param AuthenticationManagerInterface|null                 $authManager
+     *
+     * @throws \BadMethodCallException
+     */
+    public function __construct($tokenStorage, $accessDecisionManager, $map = null, $authManager = null)
     {
-        $this->tokenStorage = $tokenStorage;
-        $this->accessDecisionManager = $accessDecisionManager;
-        $this->map = $map;
-        $this->authManager = $authManager;
+        $oldSignature = $tokenStorage instanceof TokenStorageInterface
+            && $accessDecisionManager instanceof AccessDecisionManagerInterface
+            && $map instanceof AccessMapInterface
+            && $authManager instanceof AuthenticationManagerInterface;
+
+        $newSignature = $tokenStorage instanceof AuthorizationCheckerInterface
+            && $accessDecisionManager instanceof AccessMapInterface
+            && null === $map
+            && null === $authManager;
+
+        // confirm possible signatures
+        if (!$oldSignature && !$newSignature) {
+            throw new \BadMethodCallException('Unable to construct AccessListener, please provide the correct arguments');
+        }
+
+        if ($oldSignature) {
+            @trigger_error('The '.__CLASS__.' constructor expects 2 arguments since version X.X. Update constructor signature AccessListener(AuthorizationCheckerInterface, AccessMapInterface).', E_USER_DEPRECATED);
```

Comment:
`X.X` will be replaced with correspond version as soon as decision will be made where this code will land.